File: TypeScript/app/components/backstagepass.ts
import { Item } from "./Item";

export class BackstagePass extends Item {
  updateQuality() {
    super.updateQuality();
    if (this.sellIn < 0) {
      this.quality = 0;
    } else if (this.sellIn <= 5) {
      this.quality = Math.min(this.quality + 3, 50);
    } else if (this.sellIn <= 10) {
      this.quality = Math.min(this.quality + 2, 50);
    } else {
      this.quality = Math.min(this.quality + 1, 50);
    }
  }
}

File: Java/src/main/java/com/gildedrose/domain/BackstagePass.java
package com.gildedrose.domain;

import com.gildedrose.Item;
import com.gildedrose.service.InventoryItem;

public class BackstagePass extends InventoryItem {

    public BackstagePass(Item item) {
        setName(item.name);
        setSellIn(item.sellIn);
        setQuality(item.quality);
    }

    @Override
    public boolean qualityDecreaseInverted() {
        return true;
    }

    @Override
    public int qualityDecreaseAmount() {
        return 1;
    }

    @Override
    public int handleQualityAfterSellIn() {
        quality = 0;
        return quality;
    }

    @Override
    public int increaseQualityBelowMaximum() {
        quality = increaseQualityIfNotMaximum();

        // increase backstage pass further when sellIn date approaches
        quality = increaseBackstagePass();
        return quality;
    }

    private int increaseBackstagePass() {
        if (sellIn < 10) {
            quality = increaseQualityIfNotMaximum();

            if (sellIn < 5) {
                quality = increaseQualityIfNotMaximum();
            }
        }
        return quality;
    }

    private int increaseQualityIfNotMaximum() {
        return super.increaseQualityBelowMaximum();
    }
}
